### PR TITLE
[Core] Add Modeler to AnalysisStage

### PR DESCRIPTION
**Pull request by tteschemacher**
Hi all,

this adds the modeler to the analysis_stage. This is done in a similar way as the processes.

This requires #6605 .

Thx for the review.

### CODE REVIEW

**Comment by tteschemacher** [comment]
This PR is also ready. The only thing which can be discussed is how modelers are constructed.

File: kratos/python_scripts/analysis_stage.py [hunk 1]
```diff
@@ -1,305 +1,384 @@
...
+
+    def InitializeSolutionStep(self):
+        """This function performs all the required operations that should be executed
+        (for each step) BEFORE solving the solution step.
+        """
+        self.PrintAnalysisStageProgressInformation()
+
+        # updates, modifies and maps within the models.
+        for modeler in self._GetListOfModelers():
+            modeler.UpdateModelInitializeSolutionStep(self.model)
```
Comment (RiccardoRossi) [inline kratos/python_scripts/analysis_stage.py]: you miss the debugging echos...also this should be included in a function

File: kratos/python_scripts/analysis_stage.py [hunk 2]
```diff
@@ -1,305 +1,384 @@
...
+        """
+        raise Exception("Creation of the solver must be implemented in the derived class.")
+
+    ### Modelers
+    def _SetupGeometryModelAndModelParts(self):
+        ## Initialize Modelers
+        self._list_of_modelers = self._CreateModelers(self._GetOrderOfModelers())
+
+        # Import or generate geometry models from external input.
+        for modeler in self._GetListOfModelers():
```
Comment (RiccardoRossi) [inline kratos/python_scripts/analysis_stage.py:224]: I would have put each of this as a function.

aside the debugging echo is not included

**Review by RiccardoRossi — DISMISSED** [review]
still misses the debugging output (defaulted to off, but it should be there)

also i think we agreed of having a function for each of the loops

File: kratos/python_scripts/analysis_stage.py [hunk 3]
```diff
@@ -1,305 +1,384 @@
...
+
+    def _GetOrderOfModelers(self):
+        """ This function can be overridden in derived classes if the order of
+            the modelers matters
+        """
+        return []
+
+    def _CreateModelers(self, initialization_order):
+        """ List of modelers in following format:
+        "modelers" : {
```
Comment (philbucher) [inline kratos/python_scripts/analysis_stage.py]: the modelers should be spicified in a list in the project-params
this is the biggest flaw of the processes, but this had to be done for backward-compatibility

File: kratos/python_scripts/analysis_stage.py [hunk 4]
```diff
@@ -1,305 +1,384 @@
...
+
+    ### Modelers
+    def _GetListOfModelers(self):
+        """ This function returns the list of modelers
+        """
+        if not hasattr(self, '_list_of_modelers'):
+            raise Exception("The list of modelers was not yet created!")
+        return self._list_of_modelers
+
+    def _GetOrderOfModelers(self):
```
Comment (philbucher) [inline kratos/python_scripts/analysis_stage.py]: then you also won't need this function which is a real pain to use/understand/debug

**Review by philbucher — COMMENTED** [review]
@tteschemacher can you try overridding the `PrintData` function of the Modeler in Python and see if this shows when you `print` this object?
This would be very handy for debugging

File: kratos/python_scripts/analysis_stage.py [hunk 5]
```diff
@@ -1,305 +1,384 @@
...
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        model -- The Model to be used
+        project_parameters -- The ProjectParameters used
+        """
+        if (type(model) != KratosMultiphysics.Model):
+            raise Exception("Input is expected to be provided as a Kratos Model object")
+
+        if (type(project_parameters) != KratosMultiphysics.Parameters):
```
Comment (philbucher) [inline kratos/python_scripts/analysis_stage.py]: why do you change this?
The way it was is what codacy suggests

File: kratos/python_scripts/analysis_stage.py [hunk 6]
```diff
@@ -1,305 +1,384 @@
...
+        KratosMultiphysics.Logger.PrintInfo(self._GetSimulationName(), "STEP: ", self._GetSolver().GetComputingModelPart().ProcessInfo[KratosMultiphysics.STEP])
+        KratosMultiphysics.Logger.PrintInfo(self._GetSimulationName(), "TIME: ", self.time)
+
+    def FinalizeSolutionStep(self):
+        """This function performs all the required operations that should be executed
+        (for each step) AFTER solving the solution step.
+        """
+        # updates, modifies and maps within the models.
+        for modeler in self._GetListOfModelers():
+            modeler.UpdateModelFinalizeSolutionStep(self.model)
```
Comment (philbucher) [inline kratos/python_scripts/analysis_stage.py]: this should be removed for now right?

File: kratos/python_scripts/analysis_stage.py (hunk 5 above)
Comment (tteschemacher) [reply kratos/python_scripts/analysis_stage.py]: Yes, that as an unintended merging conflict

File: kratos/python_scripts/analysis_stage.py (hunk 6 above)
Comment (tteschemacher) [reply kratos/python_scripts/analysis_stage.py]: Yes, it is gone.

File: kratos/python_scripts/analysis_stage.py (hunk 1 above)
Comment (tteschemacher) [reply kratos/python_scripts/analysis_stage.py]: The modeler has the mEchoLevel with which the output of each modeler is controlled. Are you thinking of something additional to that?

File: kratos/python_scripts/analysis_stage.py (hunk 3 above)
Comment (philbucher) [reply kratos/python_scripts/analysis_stage.py]:
~~~js
"modelers" : [{
    "py_module" : "geom_import":
    "parameters" : {
        // settings for this modeler
    }
},{

    "py_module" : "iga_creatr_Quadrature_points",
    "parameters" : {
        // settings for this modeler
    }
}]
~~~

File: kratos/python_scripts/analysis_stage.py [hunk 7]
```diff
@@ -1,305 +1,376 @@
...
+        """Create the solver
+        """
+        raise Exception("Creation of the solver must be implemented in the derived class.")
+
+    ### Modelers
+    def _ModelersSetupGeometryModel(self):
+        # Import or generate geometry models from external input.
+        for modeler in self._GetListOfModelers():
+            if self.echo_level > 1:
+                KratosMultiphysics.Logger.PrintInfo(self._GetSimulationName(), "Modeler: ", str(modeler), " Setup Geometry Model started.")
```
Comment (tteschemacher) [inline kratos/python_scripts/analysis_stage.py:226]: @RiccardoRossi is that as you wanted it?

File: kratos/python_scripts/analysis_stage.py [hunk 8]
```diff
@@ -1,305 +1,376 @@
...
+    def Initialize(self):
+        """This function initializes the AnalysisStage
+        Usage: It is designed to be called ONCE, BEFORE the execution of the solution-loop
+        This function has to be implemented in deriving classes!
+        """
+        # Modelers:
+        self._CreateModelers()
+        self._ModelersSetupGeometryModel()
+        self._ModelersPrepareGeometryModel()
+        self._ModelersSetupModelPart()
```
Comment (tteschemacher) [inline kratos/python_scripts/analysis_stage.py:81]: Now all modeler calls are in separate functions. Do you agree like that @RiccardoRossi ?

File: kratos/python_scripts/modeler_factory.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                    if modeler_item.Has("kratos_module"):
+                        """Location of the modelers in Kratos; eg.:
+                        - KratosMultiphysics
+                        - KratosMultiphysics.FluidDynamicsApplication
+                        """
+                        kratos_module_name = modeler_item["kratos_module"].GetString()
+
+                        if not kratos_module_name.startswith("KratosMultiphysics"):
+                            kratos_module_name = "KratosMultiphysics." + kratos_module_name
+                        modeler_name = kratos_module_name + "." + modeler_name
```
Comment (philbucher) [inline kratos/python_scripts/modeler_factory.py]:
```suggestion
                        modeler_name = kratos_module_name + ".modelers." + modeler_name

```

Comment (philbucher) [reply kratos/python_scripts/modeler_factory.py]: this way the modellers would be placed in `kratos/python_scripts/modelers/` thus bringing more structure to this folder

File: kratos/python_scripts/analysis_stage.py [hunk 10]
```diff
@@ -1,305 +1,369 @@
...
+        """ List of modelers in following format:
+        "modelers" : [{
+            "modeler_name" : "geometry_import":
+            "parameters" : {
+                "echo_level" : 0:
+                // settings for this modeler
+            }
+        },{ ... }]
+        The order of intialization is specified by the
+        order within this list.
```
Comment (philbucher) [inline kratos/python_scripts/analysis_stage.py]:
```suggestion
```
no longer necessary

**Review by philbucher — COMMENTED** [review]
except one minor comment looks good to me now, nice work! :+1
afterwards I will approve

Regarding the debugging functionalities requrested by @RiccardoRossi I would say now we have a basic version that can be refined if necessary
But for this I would say we need some practical experience

ping @RiccardoRossi

File: kratos/python_scripts/modeler_factory.py (hunk 9 above)
Comment (tteschemacher) [reply kratos/python_scripts/modeler_factory.py]: I agree, that would mean that also applications need to follow this.

File: kratos/python_scripts/analysis_stage.py (hunk 10 above)
Comment (tteschemacher) [reply kratos/python_scripts/analysis_stage.py]: sounds good to me!

**Review by philbucher — DISMISSED** [review]
Awesome!

File: kratos/python_scripts/analysis_stage.py (hunk 8 above)
Comment (RiccardoRossi) [reply kratos/python_scripts/analysis_stage.py:81]: ok

File: kratos/python_scripts/analysis_stage.py (hunk 7 above)
Comment (RiccardoRossi) [reply kratos/python_scripts/analysis_stage.py:226]: in principle yes, but how does the output look like?

**Review by RiccardoRossi — COMMENTED** [review]
i am ok in principle. how does the output look like?

**Comment by RiccardoRossi** [comment]
also, how do you set the echo_level for the analysis_stage?

**Comment by philbucher** [comment]
> also, how do you set the echo_level for the analysis_stage?

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/analysis_stage.py#L34

**Comment by tteschemacher** [comment]
> i am ok in principle. how does the output look like?

It is e.g.:
```
::[KSM Simulation]:: : Modeler:  CadIoModeler
  Setup Geometry Model started.
::[KSM Simulation]:: : Modeler:  CadIoModeler
  Setup Geometry Model finished.
::[KSM Simulation]:: : Modeler:  IgaModeler
  Setup Geometry Model started.
::[KSM Simulation]:: : Modeler:  IgaModeler
  Setup Geometry Model finished.
```

Problematically about this is this is written in any case, even if the specific modeler does not has this function. But each modeler has its own echo level. For those outputs modelers should care itself.

**Comment by RiccardoRossi** [comment]
> > i am ok in principle. how does the output look like?
> 
> It is e.g.:
> 
> ```
> ::[KSM Simulation]:: : Modeler:  CadIoModeler
>   Setup Geometry Model started.
> ::[KSM Simulation]:: : Modeler:  CadIoModeler
>   Setup Geometry Model finished.
> ::[KSM Simulation]:: : Modeler:  IgaModeler
>   Setup Geometry Model started.
> ::[KSM Simulation]:: : Modeler:  IgaModeler
>   Setup Geometry Model finished.
> ```
> 
> Problematically about this is this is written in any case, even if the specific modeler does not has this function. But each modeler has its own echo level. For those outputs modelers should care itself.

ok this looks good to me. and in debug mode i don't mind that extra output

File: kratos/python_scripts/analysis_stage.py (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/python_scripts/analysis_stage.py]: ok

File: kratos/python_scripts/analysis_stage.py (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/python_scripts/analysis_stage.py]: i think this is still relevant, as well as the previous comment

**Comment by philbucher** [comment]
@RiccardoRossi you are blocking

**Comment by RiccardoRossi** [comment]
i am not any longer blocking, but the input is still to be changed according to your comment on the list of modelers. also the order stuff is yet to be removed

**Review by RiccardoRossi — COMMENTED** [review]
ok, i did not publish my comments, here they go

**Comment by philbucher** [comment]
> i am not any longer blocking, but the input is still to be changed according to your comment on the list of modelers. also the order stuff is yet to be removed

that was already done, can you please check the latest version?
now it is just the list of modelers

**Review by RiccardoRossi — APPROVED** [review]

**Comment by tteschemacher** [comment]
Thank you @philbucher, @RiccardoRossi and @KratosMultiphysics/technical-committee  for the review!

**Pull request merged**